source/plugins/go/src/log_loss_monitoring.go

```diff
@@ -0,0 +1,303 @@
+	// var current_iteration_count int64 = 0
+	num_containers_on_disk = 0
+
+	// Shared objects
+	deleted_containers_query = make(chan string, 5)    // There might be a lot more than 5 containers  of 10 chosen arbitrairly
+	deleted_containers_response = make(chan string, 5) // capacity of 10 chosen arbitrairly
+
+	process_logs_mut = sync.Mutex{}
+	read_disk_mut = sync.Mutex{}
+
+	disabled_namespaces = make(map[string]bool)
+}
+
+func SetupLogLossTracker() {
+	enabled = os.Getenv("CONTROLLER_TYPE") == "DaemonSet"
+
+	enabled = enabled && !(strings.ToLower(os.Getenv("IN_UNIT_TEST")) == "true")
+
+	// toggle env var is meant to be set by Microsoft, customer can set AZMON_DISABLE_LOG_LOSS_TRACKING through a configmap
+	enabled = enabled && (os.Getenv("AZMON_DISABLE_LOG_LOSS_TRACKING") != "true") && (os.Getenv("AZMON_DISABLE_LOG_LOSS_TRACKING_TOGGLE") != "true")
+
+	// don't count logs if stdout or stderr log collection is globally disabled
+	enabled = enabled && strings.ToLower(os.Getenv("AZMON_COLLECT_STDOUT_LOGS")) == "true" && strings.ToLower(os.Getenv("AZMON_COLLECT_STDERR_LOGS")) == "true"
+
+	if enabled {
+		log_loss_logger = createLogger("", "container-log-counts.log")
+
+		for _, excluded_namespace := range strings.Split(os.Getenv("AZMON_STDERR_EXCLUDED_NAMESPACES"), ",") {
+			disabled_namespaces[excluded_namespace] = true
+		}
+
+		for _, excluded_namespace := range strings.Split(os.Getenv("AZMON_STDOUT_EXCLUDED_NAMESPACES"), ",") {
+			disabled_namespaces[excluded_namespace] = true
+		}
+
+		write_counts_ticker := time.NewTicker(5 * time.Minute)
+		go write_telemetry(write_counts_ticker.C)
+
+		//TODO: turn this frequency down
+		track_rotations_ticker := time.NewTicker(10 * time.Second)
+		go track_log_rotations(track_rotations_ticker.C, "/var/log/pods")
+	}
+}
+
+// TODO: when scale testing also measure disk usage (should measure cpu, mem, disk, and lost logs when scale testing)
+// 			does running a separate process impact cpu/mem/disk/lost logs
+func Process_log(containerID *string, k8sNamespace *string, k8sPodName *string, containerName *string, logEntry *string, logTime *string) {
+	if enabled {
+		identifier := *k8sNamespace + "_" + *k8sPodName + "_" + *containerName
+		log_count_index, container_new := m_bytes_logged_storage.get(identifier)
+
+		if container_new {
+			// this branch only executes when a new container is seen, so it doesn't have to be as optimized
+			process_logs_mut.Lock() // only grab this lock when a new container is seen. Slices are thread safe so we don't need to lock it
+			defer process_logs_mut.Unlock()
+
+			// do garbage collection
+			if float32(atomic.LoadInt32(&num_containers_on_disk)) > float32(m_bytes_logged_storage.len())*3 {
+				// send some containerIDs to the FW thread to check if they are deleted from disk
+			write_to_chan:
+				for i := 0; i < 5; i++ {
+					deletion_query_index := (m_deletion_query_index + 1) % m_bytes_logged_storage.len()
```

Comment:
I see, m_deletion_query_index value is 0 in all instances, which means deletion_query_index always same,  then how do you ensure your picking random containerids?

Reply:
That's another embarrasing bug, there's now a unit test for it